src/Geometry/BoundingBox.h

```diff
@@ -1,43 +1,187 @@
 #pragma once
 
+// Local headers
 #include "Vector2.h"
 #include "Size.h"
 
+// Standard library headers
+#include <cmath>
+
 namespace Geometry {
-	template<class T>
-	struct BoundingBox {
-		Vector2<T> start;
-		Vector2<T> end;
-
-		T width() const;
-
-		T height() const;
-
-		Size<T> size() const;
-
-		bool contains(const Vector2<T> &point) const;
-	};
-
-	template<class T>
-	T BoundingBox<T>::width() const {
-		return end.x - start.x;
-	}
-
-	template<class T>
-	T BoundingBox<T>::height() const {
-		return end.y - start.y;
-	}
-
-	template<class T>
-	bool BoundingBox<T>::contains(const Vector2<T> &point) const {
-		return point.x >= start.x
-		       && point.x < end.x
-		       && point.y >= start.y
-		       && point.y < end.y;
-	}
-
-	template<class T>
-	Size<T> BoundingBox<T>::size() const {
-		return Size<T>(this->width(), this->height());
-	}
+  /// General class for representing a 2D maximum point
+  template<typename T>
+  class Maximum2 : public Vector2<T>
```

Comment:
Maximum of what?  My instant reading was (as we're talking about vectors) this was a vector with maximum length (x*x + y*y).  Actually its the top-right corner of a bounding box (and Minimum is the bottom-left).  So I wonder if there are better names.

Reply:
Ah I wonder if the ultimate issue is that `Vector` isn't being used as a `Vector` here its being used as a `Position`.  A vector is a direction and distance, it doesn't specify a start or end point.  Whereas, what we're doing is specifying a bounding box with a particular place in my for (0, 0).

Vectors and Positions mostly have a similar interface so the fix may be: `template<typename T> using Position<T> = Vector<T>;` or some such.

Reply:
It is a point so a vector offset from the origin (0, 0). Alternatively I am used to dealing with homogeneous points (x, y, z, w) for which a position is of the form (x, y, z, 1) and a displacement is (x, y, z, 0) so that when it is transformed by a 4 by 4 matrix a position will be rotated and translated whereas a displacement will only be rotated.

I'll try and make it a bit clearer in the code.

Reply:
I've reviewed this now and think that the comments make it clear that the vectors are not displacement vectors but points.